fdo.go

```diff
@@ -0,0 +1,264 @@
+package desktop
+
+import (
+	"bufio"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"fyne.io/fyne"
+)
+
+var iconExtensions = []string{".png", ".svg", ".xpm"}
+
+//FdoDesktop contains the information from Linux .desktop files
+type FdoDesktop struct {
+	Name     string
+	IconName string
+	IconPath string
+	Exec     string
+}
+
+//FdoResourceFormat turns a string into a usable name for fyne.Resources
+func FdoResourceFormat(name string) string {
+	str := strings.Replace(name, "-", "", -1)
+	return strings.Replace(str, "_", "", -1)
+}
+
+//FdoLookupXdgDataDirs returns a string slice of all XDG_DATA_DIRS
+func FdoLookupXdgDataDirs() []string {
+	dataLocation := os.Getenv("XDG_DATA_DIRS")
+	locationLookup := strings.Split(dataLocation, ":")
+	if len(locationLookup) == 1 && locationLookup[0] == dataLocation {
+		var fallbackLocations []string
+		fallbackLocations = append(fallbackLocations, "/usr/local/share")
+		fallbackLocations = append(fallbackLocations, "/usr/share")
+		return fallbackLocations
+	}
+	return locationLookup
+}
+
+//FdoLookupApplication looks up an application by name and returns a FdoDesktop
+func FdoLookupApplication(appName string) *FdoDesktop {
+	locationLookup := FdoLookupXdgDataDirs()
+	for _, dataDir := range locationLookup {
+		testLocation := filepath.Join(dataDir, "applications", appName+".desktop")
+		if _, err := os.Stat(testLocation); os.IsNotExist(err) {
+			continue
+		} else {
+			return NewFdoDesktop(testLocation)
+		}
+	}
+	return nil
+}
+
+//FdoLookupApplicationWinInfo looks up an application based on window info
+func FdoLookupApplicationWinInfo(title string, classes []string, command string, iconName string) *FdoDesktop {
+	desktop := FdoLookupApplication(title)
+	if desktop != nil {
+		return desktop
+	}
+	for _, class := range classes {
+		desktop := FdoLookupApplication(class)
+		if desktop != nil {
+			return desktop
+		}
+	}
+	desktop = FdoLookupApplication(command)
+	if desktop != nil {
+		return desktop
+	}
+	desktop = FdoLookupApplication(iconName)
+	return desktop
+}
+
+func reverseWalkDirectoryMatch(directory string, joiner string, iconName string) string {
+	//If the requested icon wasn't found in the Example 32x32 or scalable dirs of the theme, try other sizes within theme
+	files, err := ioutil.ReadDir(directory)
+	if err != nil {
+		fyne.LogError("", err)
```

Comment:
Can we provide an error string here (even just "Icon directory read error")? makes debugging easier